kyuubi-server/src/test/scala/org/apache/kyuubi/credentials/HadoopCredentialsManagerSuite.scala

```diff
@@ -0,0 +1,205 @@
+    val manager = new HadoopCredentialsManager()
+    manager.initialize(kyuubiConf)
+    assert(!manager.isProviderLoaded("unstable"))
+  }
+
+  test("schedule credentials renewal") {
+    val kyuubiConf = new KyuubiConf(false)
+      .set(KyuubiConf.CREDENTIALS_RENEWAL_INTERVAL, 1000L)
+    withStartedManager(kyuubiConf) { manager =>
+      val userRef = manager.getUserRef(appUser)
+      // Tolerate 100 ms delay
+      eventually(timeout(1100.milliseconds), interval(100.milliseconds)) {
+        assert(userRef.getEpoch == 1)
+      }
+    }
+  }
+
+  test("schedule credentials renewal retry when failed") {
+    val kyuubiConf = new KyuubiConf(false)
+      .set(KyuubiConf.CREDENTIALS_RENEWAL_INTERVAL, 1000L)
+      .set(KyuubiConf.CREDENTIALS_RENEWAL_RETRY_WAIT, 1000L)
+    withStartedManager(kyuubiConf) { manager =>
+      try {
+        UnstableDelegationTokenProvider.throwException = true
+
+        val userRef = manager.getUserRef(appUser)
+        // Tolerate 100 ms delay
+        eventually(timeout(2100.milliseconds), interval(100.milliseconds)) {
+          // 1 scheduled call and 2 scheduled retrying call
+          assert(UnstableDelegationTokenProvider.exceptionCount == 3)
+        }
+        assert(userRef.getEpoch == -1)
+      } finally {
+        UnstableDelegationTokenProvider.throwException = false
+      }
+    }
+  }
+
+  test("send credentials if needed") {
+    val kyuubiConf = new KyuubiConf(false)
+      .set(KyuubiConf.CREDENTIALS_RENEWAL_INTERVAL, 1000L)
+    withStartedManager(kyuubiConf) { manager =>
+      // Trigger UserCredentialsRef's initialization
+      val userRef = manager.getUserRef(appUser)
+      eventually(interval(100.milliseconds)) {
+        assert(userRef.getEpoch == 0)
+      }
+
+      val succeed = manager.sendCredentialsIfNeeded(engineId, appUser, send)
+      assert(succeed)
+
+      val engineRef = manager.getEngineRef(engineId)
+      assert(engineRef.getEpoch == userRef.getEpoch)
+    }
+  }
+
+  test("credentials sending failure") {
+    withStartedManager(new KyuubiConf(false)) { manager =>
+      // Trigger UserCredentialsRef's initialization
+      val userRef = manager.getUserRef(appUser)
+      eventually(interval(100.milliseconds)) {
+        assert(userRef.getEpoch == 0)
+      }
+
+      val thrown = intercept[KyuubiException] {
+        manager.sendCredentialsIfNeeded(engineId, appUser, _ => throw new IOException)
+      }
+
+      assert(thrown.isInstanceOf[KyuubiException])
+      assert(thrown.getMessage == s"Failed to send new credentials to SQL engine $engineId")
+    }
+  }
+}
+
+private class ExceptionThrowingDelegationTokenProvider extends HadoopDelegationTokenProvider {
+  ExceptionThrowingDelegationTokenProvider.constructed = true
+  throw new IllegalArgumentException
+
+  override def serviceName: String = "throw"
+
+  override def delegationTokensRequired(
+      hadoopConf: Configuration,
+      kyuubiConf: KyuubiConf): Boolean =
+    throw new IllegalArgumentException
```

Comment:
how about we add some message here?

Reply:
Test class copied from Spark. Remove it since it is confusing and has no influence in test case.